feat(agents): Add on_llm_start and on_llm_end Lifecycle Hooks

uzair330 opened the pull request:
### Motivation

Currently, the `AgentHooks` provide valuable lifecycle events for the start/end of an agent run and for tool execution (`on_tool_start`/`on_tool_end`). However, developers lack the ability to observe the agent's execution at the language model level.

This PR introduces two new hooks, `on_llm_start` and `on_llm_end`, to provide this deeper level of observability. This change enables several key use cases:
-   **Performance Monitoring:** Precisely measure the latency of LLM calls.
-   **Debugging & Logging:** Log the exact prompts sent to and raw responses received from the model.
-   **Implementing Custom Logic:** Trigger actions (e.g., updating a UI, saving state) immediately before or after the agent "thinks."

### Summary of Changes

This is a focused contribution with the following changes:

-   **`src/agents/lifecycle.py`**: Added two new async method definitions, `on_llm_start` and `on_llm_end`, to the `AgentHooks` base class. The naming convention aligns with the existing `on_..._start`/`on_..._end` pattern.
-   **`src/agents/run.py`**: Inserted calls to these new hooks within `AgentRunner._get_new_response`, directly wrapping the `model.get_response()` call to ensure accurate timing and data capture.
-   **`tests/test_agent_llm_hooks.py`**: Added a new unit test file that validates the new functionality. It uses a mock model and a spy hook class to verify:
    -   The correct hook sequence for a conversational turn.
    -   The correct sequence for a tool-using turn, ensuring the new LLM hooks fire correctly around the existing tool hooks.
    -   That the system runs without error when `agent.hooks` is `None`.

### Usage Example

Here is how a developer could use the new hooks:

```python
from agents import Agent, AgentHooks
from agents.items import ModelResponse

class LLMTrackerHooks(AgentHooks):
    async def on_llm_start(self, context, agent, system_prompt, input_items):
        print(f"Agent '{agent.name}' is calling the LLM...")

    async def on_llm_end(self, context, agent, response: ModelResponse):
        # Log token usage from the model's response
        if response.usage:
            print(f"LLM call finished. Tokens used: {response.usage.total_tokens}")

# Assign hooks to an agent instance
my_agent = Agent(
    name="MyMonitoredAgent",
    model=my_model,
    hooks=LLMTrackerHooks()
)
```
### Checklist

-   [x] My code follows the style guidelines of this project (checked with `ruff`).
-   [x] I have added tests that prove my feature works.
-   [x] All new and existing tests passed locally with my changes.

seratch commented:
When execute_tools_and_side_effects takes long, indeed these hooks would be helpful to understand the exact time spent by LLM. One quick thing I noticed is that probably this implementation doe not yet support streaming patterns.

@rm-openai do you think having these two hooks is good to go?

uzair330 commented:
---

### **New: Streaming Support for `on_llm_start` and `on_llm_end` Hooks**

This update adds full support for the `on_llm_start` and `on_llm_end` hooks in **streaming mode**, ensuring feature parity with non-streaming execution.

---

#### 1. Implementation (`src/agents/run.py`)

- Hook calls are now moved inside `_run_single_turn_streamed` to ensure they receive the correct turn-specific data.
- `on_llm_start` is triggered just before the `model.stream_response` call, with the resolved `system_prompt`.
- `on_llm_end` is triggered after the `ResponseCompletedEvent` and once the final `ModelResponse` is assembled.

---

#### 2. Unit Tests (`tests/test_agent_llm_hooks.py`)

- Added a new test: `test_streamed_hooks_for_chat_scenario`, which verifies that the hooks fire in the correct sequence using `Runner.run_streamed`.
- Confirms streaming and non-streaming execution paths are now consistent.

---

### ✅ Streaming Example

The hooks now work seamlessly with `Runner.run_streamed`, enabling real-time insight into LLM usage:

```python

import asyncio
from typing import Any

from dotenv import load_dotenv

from agents.agent import Agent
from agents.items import ItemHelpers, ModelResponse
from agents.lifecycle import AgentHooks, RunContextWrapper
from agents.run import Runner

load_dotenv()


class LLMTrackerHooks(AgentHooks[Any]):
    async def on_llm_start(
        self,
        context: RunContextWrapper,
        agent: Agent,
        system_prompt: str,
        input_items: Any,
    ) -> None:
        print(f"[HOOK] on_llm_start: LLM is starting, system_prompt={system_prompt}")

    async def on_llm_end(
        self,
        context: RunContextWrapper,
        agent: Agent,
        response: ModelResponse,
    ) -> None:
        total = response.usage.total_tokens if response.usage else None
        print(f"[HOOK] on_llm_end: LLM returned, total_tokens={total}")


async def main():
    agent = Agent(
        name="MyStreamingAgent",
        instructions="You are a helpful assistant that can answer questions and perform tasks.",
        hooks=LLMTrackerHooks(),
          )

    stream = Runner.run_streamed(agent, input="Create a Python script that prints 'Hello, World!'")
    async for event in stream.stream_events():
        if event.type == "raw_response_event":
            continue
        if event.type == "agent_updated_stream_event":
            print(f"[EVENT] agent_updated → {event.new_agent.name}")
        elif event.type == "run_item_stream_event":
            item = event.item
            if item.type == "tool_call_item":
                print("[EVENT] tool_call_item")
            elif item.type == "tool_call_output_item":
                print(f"[EVENT] tool_call_output_item → {item.output}")
            elif item.type == "message_output_item":
                text = ItemHelpers.text_message_output(item)
                print(f"[EVENT] message_output_item → {text}")

    print(f"\n[RUN COMPLETE] final_output → {stream.final_output}")


if __name__ == "__main__":
    asyncio.run(main())

```

uzair330 commented:
Hi @seratch ,
Just wanted to gently follow up on this PR and the streaming support follow-up PR. It’s been around ten days since the last update. Please let me know if there’s anything else I can do to help move forward—happy to rebase, split further, or clarify any details. Thanks again for your time!

uzair330 commented:
This PR is still relevant and ready for review.  
I’ve synced it with the latest `main`, verified all CI checks and tests pass locally, and confirmed the new hooks work as expected.

@seratch  @rm-openai  — would appreciate a quick look or approval when you have a moment. Thanks!

seratch reviewed (approved):
To me, these changes are safe and this is a good enhancement for certain use cases.

uzair330 commented:
Hello @seratch and @rm-openai,

Thanks for the reviews and approvals. I’ve rebased this branch onto the latest main, resolved all conflicts, and confirmed it’s up to date. CI jobs are currently pending and appear to require maintainer approval to run. The branch is ready for final verification — please trigger the workflows when convenient, or let me know if you’d like any further changes.
